src/gui/tray/UserModel.cpp

```diff
@@ -496,11 +499,26 @@ void User::slotItemCompleted(const QString &folder, const SyncFileItemPtr &item)
             _activityModel->addIgnoredFileToList(activity);
         } else {
             // add 'protocol error' to activity list
+            if (item->_status == SyncFileItem::Status::FileNameInvalid) {
+                showOsNotification(item->_file, activity._subject);
+            }
             _activityModel->addErrorToActivityList(activity);
         }
     }
 }
 
+int User::notificationId(const QString &subject, const QString &message)
+{
+    const QString idKey = subject + message;
+    const auto idsIter = _notificationIds.find(idKey);
+    if (idsIter != _notificationIds.end()) {
+        return *idsIter;
+    }
+    const auto newId = _notificationIds.size();
+    _notificationIds[idKey] = newId;
```

Comment:
Hm... this will grow "infinitely" over time. At least as long as you get variations on subject and message. Since it contains full concatenation of both... Also said id is kind of fragile and will break as long as you put in place strategies to remove items from that QHash (since you'll start recycling ids).

It needs to be rethought I think. For a start, why did you drop the id member of the activity object? I'd presume it was a better id to start with? Or is the issue that it changes all the time?

If really you need a way to compute more robust numeric ids out of subject and message, I'd advise using the qHash function on this idKey, then you probably don't even need to store it. That raises the question on how often that'd collide though since it's a qint16 if my memory serves me well. I'd really keep that as last resort option though if the id field in the activity objects somehow don't cut it.

Reply:
The reason why I don't use the activity id is that this id comes from the json response of the activities that the server sent to the client. The file name warnings have obviously no such id. So I need to generate a artificial id for them. Also I need to make sure that the ids of the activities and the ids of the file name warnings do not collide. Original this whole notification cache mechanism was only used for the activities that the server sent so it was fine to use the activity id for notification caching.

I'm not that satisfied with that solution too.